pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/JSymbolTable.java

```diff
@@ -0,0 +1,121 @@
+ *      <li>If there is none, it asks the same question to its parent table recursively
+ *          and returns that result.
+ * </ul>
+ * This allows directly encoding shadowing and hiding mechanisms in the parent-child
+ * relationships.
+ *
+ * <h2>Why not keep the current symbol table</h2>
+ *
+ * <p>The current symbol table framework was not built with the same goals in mind.
+ * It indexes the AST to reduce it to a simpler representation, which is mostly
+ * shortcuts to nodes. That representation hasn't proved very useful in rules,
+ * which mostly only use it to resolve variable accesses.
+ *
+ * <p>The biggest issue is that it was not designed to abstract over whether we
+ * have a node to represent a declaration or not. It can't work on reflection
+ * data, and thus cannot really help type resolution, even if a good symbol table
+ * would take the burden of resolving references off a type checker. The
+ * shortcomings of the current symbol table make the current typeres duplicate
+ * logic, and ultimately perform tasks that are not its responsibility,
+ * which is probably why {@link ClassTypeResolver} is so huge and nasty.
+ *
+ * <p>Having an abstraction layer to unify them allows the AST analyses to
+ * be complementary, and rely on each other, instead of being so self-reliant.
+ * The abstraction provided by {@link JElementSymbol} may in the future be used
+ * to build global indices of analysed projects to implement multifile analysis.
+ *
+ * <p>The goals of this rewrite should be:
+ * <ul>
+ *      <li>To make our language analyses passes share information and be complementary
+ *      <li>To have a symbol table that is precise and exhaustive enough that all rules can depend on it
+ *      <li>To modularize the system so that it's more easily testable and documentable
+ * </ul>
+ *
+ * @author Clément Fournier
+ * @since 7.0.0
+ */
+// @formatter:on
+@Experimental
+public interface JSymbolTable {
+
+    /**
+     * Returns the parent of this table, that is, the symbol table that will be
+     * delegated to if this table doesn't find a declaration.
+     *
+     * @return a symbol table, or null if this is the top-level symbol table
+     */
+    JSymbolTable getParent();
+
+
+    /**
+     * Resolves the type referred to by the given name. This must be a simple name,
+     * ie, parameterized types and array types are not available. Primitive types are
+     * also not considered because it's probably not useful.
+     *
+     * <p>The returned type reference may be a {@link JResolvableClassSymbol}, a
+     * {@link JClassSymbol} if the type was already resolved, or a {@link JTypeParameterSymbol}.
+     *
+     * <p>Doesn't handle primitive types (there's no symbol for them).
+     *
+     * @param simpleName Simple name of the type to look for
+     *
+     * @return The type reference if it can be found, otherwise an empty optional
+     */
+    Optional<? extends JSimpleTypeSymbol> resolveTypeName(String simpleName);
```

Comment:
for consistency this should be `@Nullable` instead of `Optional` now

Reply:
Thanks